src/Parallel/Invoke.hpp

```diff
@@ -0,0 +1,64 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <utility>
+
+namespace Parallel {
+// @{
+/*!
+ * \ingroup ParallelGroup
+ * \brief Send the data `args...` to the algorith running on `proxy`, and tag
+ * the message with the identifier `temporal_id`.
+ *
+ * If the algorithm was previously disabled, set `enable_if_disabled` to true to
+ * enable the algorithm on the parallel component.
+ *
+ * \note The reason there are two separate functions is because Charm++ does not
+ * allow defaulted arguments for group and nodegroup chares.
+ */
+template <typename ReceiveTag, typename Proxy, typename ReceiveDataType>
+void receive_data(Proxy&& proxy, typename ReceiveTag::temporal_id temporal_id,
+                  ReceiveDataType&& receive_data,
+                  const bool enable_if_disabled) noexcept {
+  proxy.template receive_data<ReceiveTag>(
+      std::move(temporal_id), std::forward<ReceiveDataType>(receive_data),
+      enable_if_disabled);
+}
+
+template <typename ReceiveTag, typename Proxy, typename ReceiveDataType>
+void receive_data(Proxy&& proxy, typename ReceiveTag::temporal_id temporal_id,
+                  ReceiveDataType&& receive_data) noexcept {
+  proxy.template receive_data<ReceiveTag>(
+      std::move(temporal_id), std::forward<ReceiveDataType>(receive_data));
+}
+// @}
+
+// @{
+/*!
+ * \ingroup ParallelGroup
+ * \brief Invoke a simple action on `proxy`
+ */
+template <typename Action, typename Proxy>
+void simple_action(Proxy&& proxy) noexcept {
+  proxy.template simple_action<Action>();
+}
+
+template <typename Action, typename Proxy, typename Arg0, typename... Args>
+void simple_action(Proxy&& proxy, Arg0&& arg0, Args&&... args) noexcept {
+  proxy.template simple_action<Action>(
+      std::make_tuple(std::forward<Arg0>(arg0), std::forward<Args>(args)...));
+}
+// @}
+
+/*!
+ * \ingroup ParallelGroup
+ * \brief Invoke a threaded action on `proxy`, where the proxy must be a
+ * nodegroup.
+ */
+template <typename Action, typename Proxy, typename... Args>
+void threaded_action(Proxy&& proxy, Args&&... args) noexcept {
+  proxy.template threaded_action<Action>(std::forward<Args>(args)...);
```

Comment:
`static_asset(is_node_group_proxy<Proxy>, "Proxy must be a nodegroup.");`

Reply:
Doesn't work, it could be a local object that is a nodegroup, which is the more common use case anyway.

Reply:
In that case, can't we get the proxy type from thisProxy?

Reply:
Sure, but does that work if you're passed a proxy? Is in, do proxies have `thisProxy` methods? I don't really want a bunch of code for a static_assert that would show up with a SFINAE error anyway.

Reply:
Is the SFINAE error clear?  If so that is fine; otherwise our philosophy has been to provide a clearer error message if possible.

Reply:
Since "clear error" is subjective, here is what the error is:

```
In file included from ../tests/Unit/Parallel/Test_AlgorithmNodelock.cpp:22:
../src/Parallel/Invoke.hpp:63:18: error: no member named 'threaded_action' in 'CProxy_AlgorithmArray<ArrayParallelComponent<TestMetavariables>, TestMetavariables, brigand::list<>, int, db::DataBox<brigand::list<> > >'
  proxy.template threaded_action<Action>(std::forward<Args>(args)...);
  ~~~~~          ^
../tests/Unit/Parallel/Test_AlgorithmNodelock.cpp:208:15: note: in instantiation of function template specialization 'Parallel::threaded_action<nodegroup_threaded_receive, CProxy_AlgorithmArray<ArrayParallelComponent<TestMetavariables>, TestMetavariables, brigand::list<>, int, db::DataBox<brigand::list<> > > &, const int &>' requested here
    Parallel::threaded_action<nodegroup_threaded_receive>(
              ^
../src/Parallel/Algorithm.hpp:593:21: note: in instantiation of function template specialization 'Parallel::Algorithm_detail::apply_visitor<reduce_to_nodegroup, db::DataBox<brigand::list<> >, db::DataBox<brigand::list<> >, db::DataBox<brigand::list<> > , tuples::TaggedTuple<> &, Parallel::ConstGlobalCache<TestMetavariables> &, const int &, brigand::list<>, ArrayParallelComponent<TestMetavariables> *>' requested here
  Algorithm_detail::apply_visitor<Action, InitialDataBox>(
                    ^
src/Parallel/Algorithms/AlgorithmArray.def.h:626:22: note: in instantiation of function template specialization 'Parallel::AlgorithmImpl<ArrayParallelComponent<TestMetavariables>, Parallel::Algorithms::Array, TestMetavariables, brigand::list<>, int, db::DataBox<brigand::list<> > >::simple_action<reduce_to_nodegroup>' requested here
  impl_obj->template simple_action< Action >();
                     ^
src/Parallel/Algorithms/AlgorithmArray.def.h:616:7: note: in instantiation of function template specialization 'CkIndex_AlgorithmArray<ArrayParallelComponent<TestMetavariables>, TestMetavariables, brigand::list<>, int, db::DataBox<brigand::list<> > >::_call_simple_action_void<reduce_to_nodegroup>' requested here
      _call_simple_action_void< Action >, 0, __idx, 0);
      ^
src/Parallel/Algorithms/AlgorithmArray.decl.h:96:26: note: in instantiation of function template specialization 'CkIndex_AlgorithmArray<ArrayParallelComponent<TestMetavariables>, TestMetavariables, brigand::list<>, int, db::DataBox<brigand::list<> > >::reg_simple_action_void<reduce_to_nodegroup>' requested here
      static int epidx = reg_simple_action_void< Action >();
                         ^
src/Parallel/Algorithms/AlgorithmArray.def.h:608:152: note: in instantiation of function template specialization 'CkIndex_AlgorithmArray<ArrayParallelComponent<TestMetavariables>, TestMetavariables, brigand::list<>, int, db::DataBox<brigand::list<> > >::idx_simple_action_void<reduce_to_nodegroup>' requested here
  ckBroadcast(impl_amsg, CkIndex_AlgorithmArray < ParallelComponent, Metavariables, OrderedActionsList, SpectreArrayIndex, InitialDataBox > ::template idx_simple_action_void< Action >(),0);
                                                                                                                                                       ^
../src/Parallel/Invoke.hpp:45:18: note: in instantiation of function template specialization 'CProxy_AlgorithmArray<ArrayParallelComponent<TestMetavariables>, TestMetavariables, brigand::list<>, int, db::DataBox<brigand::list<> > >::simple_action<reduce_to_nodegroup>' requested here
  proxy.template simple_action<Action>();
                 ^
../tests/Unit/Parallel/Test_AlgorithmNodelock.cpp:279:17: note: in instantiation of function template specialization 'Parallel::simple_action<reduce_to_nodegroup, CProxy_AlgorithmArray<ArrayParallelComponent<TestMetavariables>, TestMetavariables, brigand::list<>, int, db::DataBox<brigand::list<> > > &>' requested here
      Parallel::simple_action<reduce_to_nodegroup>(array_proxy);
                ^
../src/Parallel/Main.hpp:327:52: note: in instantiation of member function 'ArrayParallelComponent<TestMetavariables>::execute_next_global_actions' requested here
    tmpl::type_from<decltype(parallel_component)>::execute_next_global_actions(
                                                   ^
/home/nils/Research/spack/opt/spack/linux-antergos17-x86_64/gcc-6.3.1/brigand-master-pcaoburzhlfibhh4mvmcia2e6dkbry5x/include/brigand/algorithms/for_each.hpp:19:14: note: in instantiation of function template specialization 'brigand::for_each_args<(lambda at ../src/Parallel/Main.hpp:326:34), brigand::type_<ArrayParallelComponent<TestMetavariables> >, brigand::type_<NodegroupParallelComponent<TestMetavariables> > >' requested here
      return for_each_args( f, type_<Elements>()... );
             ^
/home/nils/Research/spack/opt/spack/linux-antergos17-x86_64/gcc-6.3.1/brigand-master-pcaoburzhlfibhh4mvmcia2e6dkbry5x/include/brigand/algorithms/for_each.hpp:25:20: note: in instantiation of function template specialization 'brigand::detail::for_each_impl<list, ArrayParallelComponent<TestMetavariables>, NodegroupParallelComponent<TestMetavariables> , (lambda at ../src/Parallel/Main.hpp:326:34)>' requested here
    return detail::for_each_impl( List{}, f );
                   ^
../src/Parallel/Main.hpp:326:9: note: in instantiation of function template specialization 'brigand::for_each<brigand::list<ArrayParallelComponent<TestMetavariables>, NodegroupParallelComponent<TestMetavariables> >, (lambda at ../src/Parallel/Main.hpp:326:34)>' requested here
  tmpl::for_each<component_list>([this](auto parallel_component) noexcept {
        ^
src/Parallel/Main.def.h:183:13: note: in instantiation of member function 'Parallel::Main<TestMetavariables>::execute_next_phase' requested here
  impl_obj->execute_next_phase();
            ^
src/Parallel/Main.def.h:173:7: note: in instantiation of member function 'Parallel::CkIndex_Main<TestMetavariables>::_call_execute_next_phase_void' requested here
      _call_execute_next_phase_void, 0, __idx, 0);
      ^
src/Parallel/Main.decl.h:78:26: note: in instantiation of member function 'Parallel::CkIndex_Main<TestMetavariables>::reg_execute_next_phase_void' requested here
      static int epidx = reg_execute_next_phase_void();
                         ^
src/Parallel/Main.def.h:223:3: note: in instantiation of member function 'Parallel::CkIndex_Main<TestMetavariables>::idx_execute_next_phase_void' requested here
  idx_execute_next_phase_void();
  ^
../src/Parallel/CharmRegistration.hpp:182:14: note: in instantiation of member function 'Parallel::CkIndex_Main<TestMetavariables>::__register' requested here
    CkIndex::__register(get_template_parameters_as_string<Chare>().c_str(),
             ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/unique_ptr.h:825:34: note: in instantiation of member function 'Parallel::charmxx::RegisterChare<Parallel::Main<TestMetavariables>, Parallel::CkIndex_Main<TestMetavariables> >::register_with_charm' requested here
    { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
                                 ^
../src/Parallel/CharmRegistration.hpp:440:12: note: in instantiation of function template specialization 'std::make_unique<Parallel::charmxx::RegisterChare<Parallel::Main<TestMetavariables>, Parallel::CkIndex_Main<TestMetavariables> >>' requested here
      std::make_unique<Derived>();
           ^
../src/Parallel/CharmRegistration.hpp:456:24: note: in instantiation of function template specialization 'Parallel::charmxx::register_func_with_charm<Parallel::charmxx::RegisterChare<Parallel::Main<TestMetavariables>, Parallel::CkIndex_Main<TestMetavariables> > >' requested here
    Parallel::charmxx::register_func_with_charm<
                       ^
../src/Parallel/Main.hpp:69:60: note: in instantiation of static data member 'Parallel::charmxx::RegisterChare<Parallel::Main<TestMetavariables>, Parallel::CkIndex_Main<TestMetavariables> >::registrar' requested here
        Main<Metavariables>, CkIndex_Main<Metavariables>>::registrar;
                                                           ^
../src/Parallel/CharmMain.cpp:134:9: note: in instantiation of member function 'Parallel::Main<TestMetavariables>::~Main' requested here
  (void)charmxx_main_component{
        ^
1 error generated.
```

Reply:
okay, clear enough for me